### PR TITLE
[Enhancement] add partition scan num limit when query internal olap table

### PR DESCRIPTION
**Pull request by MatthewH00**
## Why I'm doing:
when query big size internal olap table with full table scan or scan too many partitions, would cause BE/CN node high load, lead to cluster instability.

## What I'm doing:
add a new FE session variable `scan_olap_partition_num_limit` to limit partition scan num when query internal olap table.
(default value is 0, means no limitation)

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

### CODE REVIEW

**Comment by MatthewH00** [comment]
@kevincai Hi Could you please review the pr when have free time?
the pr add a fe session variable to limit partition scan num, avoid cluster instability caused by full table scan or scan too many partitions when query big size internal olap table.

**Review by kevincai — COMMENTED** [review]
should add UT to enforce the behavior of the session variable. Sqltester is an add-on testing in case it is too complicated to cover the code path in UT.

**Comment by MatthewH00** [comment]
> should add UT to enforce the behavior of the session variable. Sqltester is an add-on testing in case it is too complicated to cover the code path in UT.

you mean set variable diffrent value like `-1,0,a,2..`?  add more diffrent value in Sqltester should be ok?

**Comment by kevincai** [comment]
> > should add UT to enforce the behavior of the session variable. Sqltester is an add-on testing in case it is too complicated to cover the code path in UT.
> 
> you mean set variable diffrent value like `-1,0,a,2..`? add more diffrent value in Sqltester should be ok?

direct Unit test cases are preferred  in fe/fe-core/

**Comment by MatthewH00** [comment]
> should add UT to enforce the behavior of the session variable. Sqltester is an add-on testing in case it is too complicated to cover the code path in UT.

@kevincai Please review again. i add ut in the pr and fix the problem you raise above.

**Review by kevincai — DISMISSED** [review]

**Comment by kevincai** [comment]
> > should add UT to enforce the behavior of the session variable. Sqltester is an add-on testing in case it is too complicated to cover the code path in UT.
> 
> @kevincai Please review again. i add ut in the pr and fix the problem you raise above.

Wondering why `[4191, 4192]` are not covered by the tests. Is the interface needed at all?

**Comment by MatthewH00** [comment]
> Wondering why `[4191, 4192]` are not covered by the tests. Is the interface needed at all?

the function `setScanOlapPartitionNumLimit[4191, 4192]` is not necessary after testing. 
since all other variables include setter function, so i retain it.

**Comment by MatthewH00** [comment]
> > Wondering why `[4191, 4192]` are not covered by the tests. Is the interface needed at all?
> 
> the function `setScanOlapPartitionNumLimit[4191, 4192]` is not necessary after testing. since all other variables include setter function, so i retain it.

@kevincai Could you please help to push the pr find other R&D to review?

For the `setScanOlapPartitionNumLimit[4191, 4192]` is necessary or not, i find other currently exists session variable to test. similarly find the setter function is not necessary(could take affect by `set xxx=value` even though without setter function), you could find related R&D to confirm it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java [hunk 1]
```diff
@@ -137,6 +156,22 @@ public static LogicalOlapScanOperator mergePartitionPrune(LogicalOlapScanOperato
...
+            scanOlapPartitionNumLimit = ConnectContext.get().getSessionVariable().getScanOlapPartitionNumLimit();
+        } catch (Exception e) {
+            LOG.warn("fail to get variable scan_olap_partition_num_limit, set default value 0, msg: {}", e.getMessage());
+        }
+        if (scanOlapPartitionNumLimit > 0 && selectedPartitionNum > scanOlapPartitionNumLimit) {
+            String msg = "Exceeded the limit of number of olap table partitions to be scanned. " +
+                         "Number of partitions allowed: " + scanOlapPartitionNumLimit +
+                         ", number of partitions to be scanned: " + selectedPartitionNum +
+                         ". Please adjust the SQL or change the limit by set variable scan_olap_partition_num_limit.";
+            throw new AnalysisException(msg);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java:171]: StarRocksPlannerException

**Review by Seaven — DISMISSED** [review]

**Review by kevincai — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java [hunk 2]
```diff
@@ -93,6 +97,13 @@ public static LogicalOlapScanOperator prunePartitions(LogicalOlapScanOperator lo
                     logicalOlapScanOperator.getColumnMetaToColRefMap(), selectedPartitionIds);
         }
 
+        try {
+            checkScanPartitionLimit(selectedPartitionIds.size());
+        } catch (StarRocksPlannerException e) {
+            LOG.warn("{} queryId: {}", e.getMessage(), DebugUtil.printId(ConnectContext.get().getQueryId()));
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java:103]: nit:
```suggestion
            LOG.warn("queryId: {}", DebugUtil.printId(ConnectContext.get().getQueryId()), e);
```

Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java:103]: avoid print exception stack whenever possible.

**Comment by MatthewH00** [comment]
@kevincai @Seaven Please review again.
`AnalysisException` has adjusted to `StarRocksPlannerException` in `checkScanPartitionLimit` funtion.

**Review by kevincai — APPROVED** [review]

**Comment by kaijianding** [comment]
@MatthewH00  I happenly proposed a similar solution to limit partition scan number, which is adding a rule in resource group.  https://github.com/StarRocks/starrocks/pull/53916   Could you check my pr? We can discuss a best solution to this idea.

I have two questions about this pr:
1. If a query is a complex join query, is this `scan_olap_partition_num_limit` applying to all the used tables in this query?
2. If I want to limit the partition scan number in production when there is heavy load, how to set the  session `scan_olap_partition_num_limit `?  There are so many incoming query sources, it's difficult to ask all users to set this session immediately. If setting a global session, then all tables in all queries are affected.

**Comment by MatthewH00** [comment]
> @MatthewH00 I happenly proposed a similar solution to limit partition scan number, which is adding a rule in resource group. #53916 Could you check my pr? We can discuss a best solution to this idea.
> 
> I have two questions about this pr:
> 
> 1. If a query is a complex join query, is this `scan_olap_partition_num_limit` applying to all the used tables in this query?
> 2. If I want to limit the partition scan number in production when there is heavy load, how to set the  session `scan_olap_partition_num_limit `?  There are so many incoming query sources, it's difficult to ask all users to set this session immediately. If setting a global session, then all tables in all queries are affected.

1.the variable `scan_olap_partition_num_limit` is suitable to per table when query is a complex join query.
2.like current variable `query_timeout\query_mem_limit` ,  the variable `scan_olap_partition_num_limit` is for cluster level , not set on resouce group.
@kevincai you could consider two pr , see which one is better

**Comment by MatthewH00** [comment]
> 1.the variable `scan_olap_partition_num_limit` is suitable to per table when query is a complex join query. 2.like current variable `query_timeout\query_mem_limit\scan_hive_partition_num_limit` , the variable `scan_olap_partition_num_limit` is for cluster level , not set on resouce group. @kevincai you could consider two pr , see which one is better

@kevincai @kaijianding I think both retain may be good. if user not set resouce group could use `scan_olap_partition_num_limit`, if user set resouce group could set limit at resouce group level. 
for current limit like `query_timeout/query_mem_limit` could both set cluster and resouce group level.

**Comment by MatthewH00** [comment]
> @kevincai @Seaven Please review again. `AnalysisException` has adjusted to `StarRocksPlannerException` in `checkScanPartitionLimit` funtion.

@Seaven Please review again when have free time.

**Comment by MatthewH00** [comment]
> @kevincai @Seaven Please review again. `AnalysisException` has adjusted to `StarRocksPlannerException` in `checkScanPartitionLimit` funtion.

@kevincai Hi Could you please help to push the PR?  it looks like nobody to review it since i update the code last week.

**Comment by kevincai** [comment]
> > @kevincai @Seaven Please review again. `AnalysisException` has adjusted to `StarRocksPlannerException` in `checkScanPartitionLimit` funtion.
> 
> @kevincai Hi Could you please help to push the PR? it looks like nobody to review it since i update the code last week.

will take care of it.

**Review by Seaven — APPROVED** [review]

**Comment by kaijianding** [comment]
> > > 1.the variable `scan_olap_partition_num_limit` is suitable to per table when query is a complex join query. 2.like current variable `query_timeout\query_mem_limit\scan_hive_partition_num_limit` , the variable `scan_olap_partition_num_limit` is for cluster level , not set on resouce group. @kevincai you could consider two pr , see which one is better
> > 
> > 
> > @kevincai @kaijianding I think both retain may be good. if user not set resouce group could use `scan_olap_partition_num_limit` at cluster level, if user set resouce group could set limit at resouce group level. for current limit like `query_timeout/query_mem_limit` could both set cluster and resouce group level.
> 
> @kaijianding I also tend to keep both. In your pr #53916 , you set the scan limit for each table, but the parameter name is partition_scan_number_limit_rule. Can the two parameters be unified? Set a scan limit that takes effect for all partitions instead of specifying it separately for each table.

I think every table should have its own limit. In a complex query, a bigger table should have smaller limit, a smaller table may not be limited at all.

**Comment by kaijianding** [comment]
> > > > > 1.the variable `scan_olap_partition_num_limit` is suitable to per table when query is a complex join query. 2.like current variable `query_timeout\query_mem_limit\scan_hive_partition_num_limit` , the variable `scan_olap_partition_num_limit` is for cluster level , not set on resouce group. @kevincai you could consider two pr , see which one is better
> > > > 
> > > > 
> > > > @kevincai @kaijianding I think both retain may be good. if user not set resouce group could use `scan_olap_partition_num_limit` at cluster level, if user set resouce group could set limit at resouce group level. for current limit like `query_timeout/query_mem_limit` could both set cluster and resouce group level.
> > > 
> > > 
> > > @kaijianding I also tend to keep both. In your pr #53916 , you set the scan limit for each table, but the parameter name is partition_scan_number_limit_rule. Can the two parameters be unified? Set a scan limit that takes effect for all partitions instead of specifying it separately for each table.
> > 
> > 
> > I think every table should have its own limit. In a complex query, a bigger table should have smaller limit, a smaller table may not be limited at all.
> 
> The row limit of large tables should be larger than that of small tables, so is it okay to use just one value directly. Setting different rules for different tables seems uncommon. Moreover, the data of the tables is dynamic. If different values are set for each table, does the user need to adjust the rules frequently? Here, if only considering the resource usage limit, it is more reasonable to set the same threshold for all tables.

This rule is to limit partition scan number, it's not row limit .

**Comment by MatthewH00** [comment]
@kevincai Code review looks have passed serval days ago. Could you help to merge it to the main branch when have free time?

**Comment by MatthewH00** [comment]
@kevincai The PR code review looks have passed last week. Could you help to merge it to the main branch when have free time?

**Comment by kaijianding** [comment]
> I think the resource group is bound to the computing resources, rather than to the table or even the partition. The partition limit seems to be applicable only when the user clearly knows the size of their table. It is difficult to set this value in scenarios where the size cannot be clearly estimated.

User can modify this rule after their table has data according to their query needs. It's easy to know the size of a table or a partition by `show data` or `show partitions`

**Pull request merged**